processing/src/main/java/org/apache/druid/query/groupby/orderby/DefaultLimitSpec.java

```diff
@@ -243,7 +246,30 @@ public int compare(Row left, Row right)
       if (postAggregatorsMap.containsKey(columnName)) {
         nextOrdering = metricOrdering(columnName, postAggregatorsMap.get(columnName).getComparator());
       } else if (aggregatorsMap.containsKey(columnName)) {
-        nextOrdering = metricOrdering(columnName, aggregatorsMap.get(columnName).getComparator());
+        Comparator comparator;
+        AggregatorFactory aggFactory = aggregatorsMap.get(columnName);
+        if (aggFactory instanceof StringLastAggregatorFactory
```

Comment:
Ouch!! We should try to avoid query engines (like groupBy) using `instanceof` to be aware of special cases for specific aggregators. It likely means something is flawed with the design somewhere.

Maybe StringLastAggregatorFactory & StringFirstAggregatorFactory should have their getComparator method changed to only look at the string value, not the timestamp. I'm not sure if there's any good reason for them to look at the timestamp.

Please add tests too. Ideally, one in TopNQueryRunnerTest and one in GroupByQueryRunnerTest.

Reply:
> Ouch!! We should try to avoid query engines (like groupBy) using instanceof to be aware of special cases for specific aggregators. It likely means something is flawed with the design somewhere.

Yes, I can't agree more. This solution is ugly.
AggregatorFactory interface provides a unified way to get comparator ("getComparator()"), but unlike other value types, strings can be compared in several ways. it seems that getComparator(orderingType) method should be more suitable for string first/last agg, but it is very confusing for other types of agg. so, a redesign may be required if this problem is to be solved more elegantly. I haven't got a very good idea yet. I hope you guys can give me some advice.

> Maybe StringLastAggregatorFactory & StringFirstAggregatorFactory should have their getComparator method changed to only look at the string value, not the timestamp. I'm not sure if there's any good reason for them to look at the timestamp.
I'm not an expert, but I also wonder why string first/last AggregatorFactory is so special.

> Please add tests too. Ideally, one in TopNQueryRunnerTest and one in GroupByQueryRunnerTest.
OK, sorry for forgetting the tests.

@gianm